File: app/src/main/java/fobo66/exchangecourcesbelarus/model/MyfinParser.java
package fobo66.exchangecourcesbelarus.model;

import android.util.Xml;
import androidx.annotation.NonNull;
import fobo66.exchangecourcesbelarus.entities.Currency;
import fobo66.exchangecourcesbelarus.util.CurrencyBuilder;
import fobo66.exchangecourcesbelarus.util.CurrencyBuilderImpl;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

/**
 * XML parser for <a href="myfin.by">MyFIN</a> feed
 * </p>
 * Created by fobo66 on 16.08.2015.
 */
public class MyfinParser implements CurrencyCourseParser {
  private static final String ns = null;
  private final List<String> neededTagNames = new ArrayList<>(
      Arrays.asList("bankname", "usd_buy", "usd_sell", "eur_buy", "eur_sell", "rur_buy",
          "rur_sell"));

  @NonNull public List<Currency> parse(@NonNull InputStream inputStream)
      throws XmlPullParserException, IOException {
    XmlPullParser parser = Xml.newPullParser();
    parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
    parser.setInput(inputStream, "utf-8");
    parser.nextTag();
    return readFeed(parser);
  }

  @NonNull private List<Currency> readFeed(@NonNull XmlPullParser parser)
      throws XmlPullParserException, IOException {
    List<Currency> entries = new ArrayList<>();
    Currency entry;

    parser.require(XmlPullParser.START_TAG, ns, "root");
    while (parser.next() != XmlPullParser.END_TAG) {
      if (parser.getEventType() != XmlPullParser.START_TAG) {
        continue;
      }
      String name = parser.getName();
      if (name.equals("bank")) {
        entry = readEntry(parser);
        if (entry != null) {
          entries.add(entry);
        }
      } else {
        skip(parser);
      }
    }
    return entries;
  }

  @NonNull private Currency readEntry(@NonNull XmlPullParser parser)
      throws XmlPullParserException, IOException {
    parser.require(XmlPullParser.START_TAG, ns, "bank");

    String fieldName;
    CurrencyBuilder currencyBuilder = new CurrencyBuilderImpl();

    while (parser.next() != XmlPullParser.END_TAG) {
      if (parser.getEventType() != XmlPullParser.START_TAG) {
        continue;
      }

      fieldName = parser.getName();
      if (isTagNeeded(fieldName)) {
        try {
          currencyBuilder = currencyBuilder.with(fieldName, readTag(parser, fieldName));
        } catch (Exception e) {
          e.printStackTrace();
        }
      } else {
        skip(parser);
      }
    }
    return currencyBuilder.build();
  }

  private boolean isTagNeeded(String tagName) {
    return neededTagNames.contains(tagName);
  }

  private String readTag(@NonNull XmlPullParser parser, String tagName)
      throws IOException, XmlPullParserException {
    parser.require(XmlPullParser.START_TAG, ns, tagName);
    String param = readText(parser);
    parser.require(XmlPullParser.END_TAG, ns, tagName);
    return param;
  }

  @NonNull private String readText(@NonNull XmlPullParser parser)
      throws IOException, XmlPullParserException {
    String result = "";
    if (parser.next() == XmlPullParser.TEXT) {
      result = parser.getText();
      parser.nextTag();
    }
    return result;
  }

  private void skip(@NonNull XmlPullParser parser) throws XmlPullParserException, IOException {
    if (parser.getEventType() != XmlPullParser.START_TAG) {
      throw new IllegalStateException();
    }
    int depth = 1;
    while (depth != 0) {
      switch (parser.next()) {
        case XmlPullParser.END_TAG:
          depth--;
          break;
        case XmlPullParser.START_TAG:
          depth++;
          break;
      }
    }
  }
}

File: app/src/main/java/fobo66/exchangecourcesbelarus/model/MyfinParser.kt
package fobo66.exchangecourcesbelarus.model

import android.util.Xml
import fobo66.exchangecourcesbelarus.entities.Currency
import fobo66.exchangecourcesbelarus.util.CurrencyBuilder
import fobo66.exchangecourcesbelarus.util.CurrencyBuilderImpl
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.io.InputStream
import javax.inject.Inject

/**
 * XML parser for [MyFIN](myfin.by) feed
 *
 * Created by fobo66 on 16.08.2015.
 */
class MyfinParser @Inject constructor() : CurrencyRatesParser {
  private val namespace: String? = null
  private val neededTagNames: List<String> = listOf(
    "bankname", "usd_buy", "usd_sell", "eur_buy", "eur_sell", "rur_buy",
    "rur_sell"
  )

  @Throws(XmlPullParserException::class, IOException::class)
  override fun parse(inputStream: InputStream): Set<Currency> {
    val parser = Xml.newPullParser().apply {
      setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
      setInput(inputStream, "utf-8")
      nextTag()
    }
    return readCurrencies(parser)
  }

  @Throws(XmlPullParserException::class, IOException::class)
  private fun readCurrencies(parser: XmlPullParser): Set<Currency> {
    val currencies = mutableSetOf<Currency>()
    var currency: Currency
    parser.require(XmlPullParser.START_TAG, namespace, ROOT_TAG_NAME)
    parser.read {
      if (parser.name == ENTRY_TAG_NAME) {
        currency = readCurrency(parser)
        currencies.add(currency)
      } else {
        skip(parser)
      }
    }
    return currencies
  }

  @Throws(XmlPullParserException::class, IOException::class)
  private fun readCurrency(parser: XmlPullParser): Currency {
    parser.require(XmlPullParser.START_TAG, namespace, ENTRY_TAG_NAME)
    var fieldName: String
    var currencyBuilder: CurrencyBuilder = CurrencyBuilderImpl()
    parser.read {
      fieldName = parser.name
      if (isTagNeeded(fieldName)) {
        currencyBuilder = currencyBuilder.with(fieldName, readTag(parser, fieldName))
      } else {
        skip(parser)
      }
    }
    return currencyBuilder.build()
  }

  private fun isTagNeeded(tagName: String): Boolean {
    return neededTagNames.contains(tagName)
  }

  @Throws(IOException::class, XmlPullParserException::class)
  private fun readTag(parser: XmlPullParser, tagName: String): String {
    parser.require(XmlPullParser.START_TAG, namespace, tagName)
    val param = readText(parser)
    parser.require(XmlPullParser.END_TAG, namespace, tagName)
    return param
  }

  @Throws(IOException::class, XmlPullParserException::class)
  private fun readText(parser: XmlPullParser): String {
    var result = ""
    if (parser.next() == XmlPullParser.TEXT) {
      result = parser.text
      parser.nextTag()
    }
    return result
  }

  @Throws(XmlPullParserException::class, IOException::class)
  private fun skip(parser: XmlPullParser) {
    check(parser.eventType == XmlPullParser.START_TAG)
    var depth = 1
    while (depth != 0) {
      when (parser.next()) {
        XmlPullParser.END_TAG -> depth--
        XmlPullParser.START_TAG -> depth++
      }
    }
  }

  private inline fun XmlPullParser.read(block: () -> Unit) {
    while (next() != XmlPullParser.END_TAG) {
      if (eventType != XmlPullParser.START_TAG) {
        continue
      }

      block()
    }
  }

  companion object {
    const val ROOT_TAG_NAME = "root"
    const val ENTRY_TAG_NAME = "bank"
  }
}